subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -124,7 +136,37 @@ private void configureArchivesAndComponent(final Project project, final JavaPlug
         runtimeElementsConfiguration.getArtifacts().add(jarArtifact);
         project.getExtensions().getByType(DefaultArtifactPublicationSet.class).addCandidate(jarArtifact);
 
-        project.getComponents().add(new JavaLibrary(jarArtifact, project.getConfigurations()));
+        final JavaCompile javaCompile = (JavaCompile) project.getTasks().getByPath(COMPILE_JAVA_TASK_NAME);
+        final ProcessResources processResources = (ProcessResources) project.getTasks().getByPath(PROCESS_RESOURCES_TASK_NAME);
+        addVariants(runtimeConfiguration, jarArtifact, javaCompile, processResources);
+        addVariants(runtimeElementsConfiguration, jarArtifact, javaCompile, processResources);
+        project.getComponents().add(new JavaLibrary(project.getConfigurations(), jarArtifact));
+    }
+
+    private void addVariants(Configuration configuration, final ArchivePublishArtifact jarArtifact, final JavaCompile javaCompile, final ProcessResources processResources) {
+        NamedDomainObjectContainer<ConfigurationVariant> runtimeVariants = configuration.getOutgoing().getVariants();
+        createVariant(runtimeVariants, "jar:" + javaCompile.getName(), jarArtifact);
```

Comment:
There's no way to create unnamed variants: the domain object container needs one.